python/erdos/__init__.py

```diff
@@ -61,85 +95,98 @@ def connect(op_type, config, read_streams, *args, **kwargs):
+    """
+    def __init__(self, py_node_handle, processes):
+        self.py_node_handle = py_node_handle
+        self.processes = processes
+
+    def shutdown(self):
+        """ Shuts down the dataflow."""
+        logger.info("Shutting down other processes")
+        for p in self.processes:
+            p.terminate()
+            p.join()
+        logger.info("Shutting down node.")
+        self.py_node_handle.shutdown_node()
+
+    def wait(self):
+        """ Waits for the completion of all the operators in the dataflow """
+        for p in self.processes:
+            p.join()
+
+
+def run(graph_filename: Optional[str] = None,
+        start_port: Optional[int] = 9000):
     """Instantiates and runs the dataflow graph.
 
     ERDOS will spawn 1 process for each python operator, and connect them via
     TCP.
 
     Args:
-        graph_filename (str): the filename to which to write the dataflow graph
+        graph_filename: The filename to which to write the dataflow graph
             as a DOT file.
-        start_port (int): the port on which to start. The start port is the
+        start_port: The port on which to start. The start port is the
             lowest port ERDOS will use to establish TCP connections between
             operators.
     """
-    data_addresses = [
-        "127.0.0.1:{port}".format(port=start_port + i)
-        for i in range(_num_py_operators + 1)
-    ]
-    control_addresses = [
-        "127.0.0.1:{port}".format(port=start_port + len(data_addresses) + i)
-        for i in range(_num_py_operators + 1)
-    ]
-
-    def runner(node_id, data_addresses, control_addresses):
-        _internal.run(node_id, data_addresses, control_addresses)
-
-    processes = [
-        mp.Process(target=runner, args=(i, data_addresses, control_addresses))
-        for i in range(1, _num_py_operators + 1)
-    ]
-
-    for p in processes:
-        p.start()
-
-    # Needed to shut down child processes
-    def sigint_handler(sig, frame):
-        for p in processes:
-            p.terminate()
-        sys.exit(0)
-
-    signal.signal(signal.SIGINT, sigint_handler)
+    driver_handle = run_async(graph_filename, start_port)
+    logger.debug("Waiting for the dataflow to complete ...")
+    driver_handle.wait()
 
-    # The driver must always be on node 0 otherwise ingest and extract streams
-    # will break
-    _internal.run_async(0, data_addresses, control_addresses, graph_filename)
 
-    for p in processes:
-        p.join()
-
-
-def run_async(graph_filename=None, start_port=9000):
+def run_async(graph_filename: Optional[str] = None,
+              start_port: Optional[int] = 9000) -> NodeHandle:
     """Instantiates and runs the dataflow graph asynchronously.
 
     ERDOS will spawn 1 process for each python operator, and connect them via
     TCP.
 
     Args:
-        graph_filename (str): the filename to which to write the dataflow graph
+        graph_filename: The filename to which to write the dataflow graph
```

Comment:
What guide lines are you following? Are we removing the types from the comments if we're using type annotations?

Reply:
[Google style guide gets rid of types in parentheses](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) when there's type hints. After discussing with @sukritkalra, we might want to keep around types in parentheses for attributes because Sphinx can't currently infer those types.